### DISCUSSION TITLE
[L11] Is there a way to disable merging with the base config?

### DISCUSSION THREAD

**Discussion by mfn** [laravel/framework#50876] [Q&A] [answered]
https://github.com/laravel/framework/pull/47309 forces now everyone to live with base configs merged in your configs with the introduction of `\Illuminate\Foundation\Bootstrap\LoadConfiguration::getBaseConfiguration` and the logic in `loadConfigurationFile()` and `mergeableOptions()`.

I understand how and why, but in my case I want full control, as I have with < Laravel 11, what is in the config and what not. It becomes risky to perform updates if updates to base configs are made and defaults might change and you might not find out before production due to different environments etc.

TL;DR: is there an easy way to disable this behaviour without replacing `LoadConfiguration` bootstrapper?

**Comment by puklipo** [comment]
### Safest solution
Publish all internal config files.
```
php artisan config:publish --all
```

## Dangerous solution
If there is a config cache, internal config will not be read.
Create your own `config:cache` command and create a cache with just the project config.
However, just caching the project config will not work.

**Reply by mfn** [reply]
Thanks, but that's not the solution I'm looking for: does not solve the problem that in uncached environemnts, read: local dev, it's still "merged".

**Reply by jonnylink** [reply]
This solution also doesn't solve for undesired config items getting added back in. I'm not sure what the idea was, but I can't see any use case for it. New config items could be part of a upgrade guide. If we've deliberately removed database and cache connections, mail services that don't exist and the like, then having them magically added back can (and in our case did) cause unexpected and undesired outcomes.

**Comment by jonnylink** [comment]
There should definitely be a way to do this. The config files having all sorts of stuff we don't want is not helpful. Database config for example should never have connections that don't exist.

**Reply by mfn** [reply]
Yes, database config is the first where I noticed it: we've a customized logic in tests to iterate over connections _we_ defined, but suddenly there were all sorts of connections present we never defined.

That's when I fell into the rabbit hole. I haven't yet tried replacing `LoadConfiguration`, if/how feasible it is. But L11 upgrade definitely took a hard break due to this.

**Reply by jonnylink** [reply]
Yeah, very similar situation here. Very frustrating. This level of "magic" seems like it'll cause more problems than it could ever solve.

**Reply by browner12** [reply]
that's my gut reaction so far. seems like a solution in search of a problem.

**Reply by taka-oyama** [reply]
> Database config for example should never have connections that don't exist

Totally agree with this...

**Comment by taka-oyama** [comment, marked as answer]
I've created a PR that will allow us to disable this feature.
Please take a look at it and give feedback. Thanks.

https://github.com/laravel/framework/pull/51579
